src/annot/spec.ml

```diff
@@ -0,0 +1,378 @@
+  | unop term_1                    ==> Unop (unop, term_1)
+  | binop term_1 term_2            ==> BinOp (binop, term_1, term_2)
+
+binpred ::= '>='                   ==> Ge
+  | '>'                            ==> Gt
+  | '<='                           ==> Le
+  | '<'                            ==> Lt
+  | '='                            ==> Eq
+  | '!='                           ==> Neq
+
+unconnect ::= '!'                  ==> Not
+
+binconnect ::= '&&'                ==> And
+  | '||'                           ==> Or
+  | '==>'                          ==> Imply
+  | '<==>'                         ==> Equiv
+
+prop ::= '(' pprop ')'             ==> pprop
+  | 'true'                         ==> Const true
+  | 'false'                        ==> Const false
+
+binder ::= 'forall'                ==> Forall
+  | 'exists'                       ==> Exists
+
+binder_type ::= 'i32'              ==> I32
+  | 'i64'                          ==> I64
+  | 'f32'                          ==> F32
+  | 'f64'                          ==> F64
+
+pprop ::= binpred term_1 term_2    ==> BinPred (binpred, term_1, term_2)
+  | unconnect prop_1               ==> UnConnect (unconnect, prop_1)
+  | binconnect prop_1 prop_2       ==> BinConnect (binconnect, prop_1, prop_2)
+  | binder binder_type prop_1      ==> Binder (binder, binder_type, None, prop_1)
+  | binder binder_type ind prop_1  ==> let* ind = (parse_text_id_result ind) in
+                                          Binder (binder, binder_type, Some ind, prop_1)
+
+*)
+
+type nonrec binpred =
+  | Ge
+  | Gt
+  | Le
+  | Lt
+  | Eq
+  | Neq
+
+type nonrec unconnect = Not
+
+type nonrec binconnect =
+  | And
+  | Or
+  | Imply
+  | Equiv
+
+type nonrec binder =
+  | Forall
+  | Exists
+
+type nonrec binder_type = num_type
+
+type nonrec unop =
+  | Neg
+  | CustomUnOp of string (* for testing purpose only *)
+
+type nonrec binop =
+  | Plus
+  | Minus
+  | Mult
+  | Div
+  | CustomBinOp of string (* for testing purpose only *)
+
+type 'a term =
+  | Int32 : Int32.t -> 'a term
+  | Int64 : Int64.t -> 'a term
+  | Float32 : Float32.t -> 'a term
+  | Float64 : Float64.t -> 'a term
+  | Var : text indice -> text term
+  | ParamVar : 'a indice -> 'a term
+  | GlobalVar : 'a indice -> 'a term
+  | BinderVar : 'a indice -> 'a term
+  | UnOp : unop * 'a term -> 'a term
+  | BinOp : binop * 'a term * 'a term -> 'a term
+  | Result : 'a term
+
+type 'a prop =
+  | Const : bool -> 'a prop
+  | BinPred : binpred * 'a term * 'a term -> 'a prop
+  | UnConnect : unconnect * 'a prop -> 'a prop
+  | BinConnect : binconnect * 'a prop * 'a prop -> 'a prop
+  | Binder : binder * binder_type * string option * 'a prop -> 'a prop
+
+let pp_bool fmt = function true -> pf fmt "true" | false -> pf fmt "false"
+
+let pp_binpred fmt = function
+  | Ge -> pf fmt "≥"
+  | Gt -> pf fmt ">"
+  | Le -> pf fmt "≤"
+  | Lt -> pf fmt "<"
+  | Eq -> pf fmt "="
+  | Neq -> pf fmt "≠"
```

Comment:
I'm fine about using the nicer `≠`, `≥` etc. But you'll have to make sure that we can parse them also. Is it the case already?

Reply:
No, I'm currently using `!=`, `>=` for syntax. I'm not sure we can parse `≠`, `≥`, etc. I'll try it

Reply:
To be clear, I mean that we should allow both when parsing, but use the nicer one when printing.

Reply:
fixed